database/database.rs

```diff
@@ -83,55 +86,94 @@ impl<D> fmt::Debug for Database<D> {
 
-    pub(super) fn reserve_write_transaction(&self, timeout_millis: u64) {
-        // TODO: TransactionError
-        let mut guard = self.schema_write_transaction_exclusivity.lock().unwrap();
+    pub(super) fn reserve_write_transaction(&self, timeout_millis: u64) -> Result<(), TransactionError> {
+        let (mut guard, timeout_left) =
+            self.try_acquire_schema_write_transaction_lock(Duration::from_millis(timeout_millis))?;
         let (has_schema_transaction, running_write_transactions, ref mut notify_queue) = *guard;
+
         if has_schema_transaction || !notify_queue.is_empty() {
             let (sender, receiver) = sync_channel::<()>(0);
             notify_queue.push_back(TransactionReservationRequest::Write(sender));
             drop(guard);
-            receiver.recv().unwrap();
-            // TODO: turn into TransactionError on timeout
+            receiver.recv_timeout(timeout_left).map_err(|source| TransactionError::Timeout { source })?;
         } else {
             guard.1 = running_write_transactions + 1;
             drop(guard);
         }
+        Ok(())
     }
 
-    pub(super) fn reserve_schema_transaction(&self, timeout_millis: u64) {
-        let mut guard = self.schema_write_transaction_exclusivity.lock().unwrap();
+    pub(super) fn reserve_schema_transaction(&self, timeout_millis: u64) -> Result<(), TransactionError> {
+        let (mut guard, timeout_left) =
+            self.try_acquire_schema_write_transaction_lock(Duration::from_millis(timeout_millis))?;
         let (has_schema_transaction, running_write_transactions, ref mut notify_queue) = *guard;
+
         if has_schema_transaction || running_write_transactions > 0 || !notify_queue.is_empty() {
             let (sender, receiver) = sync_channel::<()>(0);
             notify_queue.push_back(TransactionReservationRequest::Schema(sender));
             drop(guard);
-            receiver.recv().unwrap();
-            // TODO: turn into TransactionError on timeout
+            receiver.recv_timeout(timeout_left).map_err(|source| TransactionError::Timeout { source })?;
         } else {
             guard.0 = true;
             drop(guard);
         }
+        Ok(())
     }
 
     pub(super) fn release_write_transaction(&self) {
-        let mut guard = self.schema_write_transaction_exclusivity.lock().unwrap();
+        let mut guard = self
+            .schema_write_transaction_exclusivity
+            .lock()
+            .expect("The exclusive access should already be acquired in `reserve`");
         guard.1 -= 1;
         if guard.1 == 0 {
             Self::fulfill_reservation_requests(&mut guard)
         }
     }
 
     pub(super) fn release_schema_transaction(&self) {
-        let mut guard = self.schema_write_transaction_exclusivity.lock().unwrap();
+        let mut guard = self
+            .schema_write_transaction_exclusivity
+            .lock()
+            .expect("The exclusive access should already be acquired in `reserve`");
```

Comment:
I've also started using `expect` to make debugging these panics so much easier! We should keep using them.